datavault-common/src/main/java/org/datavaultplatform/common/model/dao/DepositDAOImpl.java

```diff
@@ -158,8 +181,43 @@ public int inProgressCount() {
     }
 
     @Override
-    public Long size() {
+    public Long size(String userId) {
         Session session = this.sessionFactory.openSession();
-        return (Long)session.createCriteria(Deposit.class).setProjection(Projections.sum("depositSize")).uniqueResult();
+        Optional<Criteria> depositCriteria = depositCriteriaForUser(userId, session, Permission.CAN_VIEW_VAULTS_SIZE);
+        if (depositCriteria.isPresent()) {
+            return (Long) depositCriteria.get().setProjection(Projections.sum("depositSize")).uniqueResult();
+        }
+        return 0L;
+    }
+
+    private Optional<Criteria> depositCriteriaForUser(String userId, Session session, Permission permission) {
```

Comment:
So what it does is:
1. Find all assigned roles from everywhere where the user has a specified permission
2. If the role has no school, we set the school id as `*`
3. If no schools were returned, we return empty to signify that the user should not be able to run this query at all
4. If any of the roles gave the `*` school id, then assume the user has access to all schools an return unrestricted criteria
5. Otherwise return criteria that include restriction for  data for those schools.

Does my understanding seem correct? If so, this is a bit... wonky. I'd probably have split that to two methods to start with, as this is called `[get]depositCriteria`, but it actually runs a query while doing it. Kind of against the "side effects" principle. 

I'm (from the top of my head) thinking about sth like:

```
Optional<Collection<String>> getSchoolIdsForUser(String userId, Session session, Permission permission)
{
  ...
  if(roleAssignment.getSchool() == null) return Optional.empty(); // * case, no entries
  ...
}

boolean applySchoolRestrictionsToCriteria(Criteria criteria, Collection<String> schoolIds)
{
  if(schoolIds.count() == 0) return false; // no schools allowed
  ... // apply criteria
}
...

Optional<Collection<String>> accessibleSchoolIds = getSchoolIdsForUser(userId, session, permission);
Criteria depositCriteria = session.createCriteria(Deposit.class, "deposit");

if(accessibleSchoolIds.isPresent())
  if(!applySchoolRestrictionsToCriteria(depositCriteria, acessibleSchoolIds.get()))
    return new ArrayList<>(); // return 0; return false;, whatever is applicable

depositCriteria.get()...
```

This is still wonky, as we are using return of the `apply...` method to signify "no access", but it's really a short-circuit for making a "return none" criteria.

Reply:
Your understanding seems correct to me.

I see what you mean about side effects - that's why I didn't preface the method name with `get` because it's doing more than just "getting" something. My main problem with it is less that it's running a query to work out how to build up the `Criteria` and more that it also checks if there's any permitted schools as part of the same method - if I could express "this person has no permissions" as a `Criteria` rather than having to introduce the `Optional` wrapper for that I don't think I'd have a problem with it.

I think maybe a better way of doing it would be to pull out as much of it as I can into it's own class that's something like:

```
class SchoolPermissionCriteriaBuilder  {
    private Set<String> schoolIds;
    public void setSchoolIds(Set<String> schoolIds) {
        this.schoolIds = schoolIds;
    }
    ...
    public boolean hasNoAccess() {
        return schoolIds.isEmpty();
    }

    public Criteria build() {
    }
}
```

Then pull the query to get the permitted school IDs into its own utils class (because having it in the builder would definitely break the single responsibility principle), and in the calling code we just do something like this:

```
SchoolPermissionCriteriaBuilder criteriaBuilder = new SchoolPermissionCriteriaBuilder();
criteriaBuilder.setSchoolIds(DaoUtils.getPermittedSchools(...));
....
if (criteriaBuilder.hasNoAccess()) {
    return new ArrayList<>();
}
Criteria criteria = criteriaBuilder.build();
...
```